services/backend/src/recommendation_system/ml_models/alpha/two_tower.py

```diff
@@ -36,54 +54,547 @@ def forward_user(self, user_tensor):
+            on="user_id",
+        )
+        del user_vector_df
+
+        # Unpack prompt embedding
+        prompt_columns = [
+            f"prompt_embedding_{i}" for i in range(PROMPT_EMBEDDING_LENGTH)
+        ]
+        df[prompt_columns] = pd.DataFrame(
+            df["prompt_embedding"].tolist(), index=df.index
+        )
+        df = df.drop("prompt_embedding", axis=1)
+
+        # Create overall features
+        user_columns = (
+            [f"ms_engaged_{i}" for i in range(TOP_CONTENT)]
+            + [f"like_vector_{i}" for i in range(TOP_CONTENT)]
+            + [f"dislike_vector_{i}" for i in range(TOP_CONTENT)]
+        )
+        user_columns.insert(0, "user_id")
+        user_features = df[user_columns]
+
+        content_columns = (
+            list(filter(lambda x: "artist_style_" in x, df.columns))
+            + list(filter(lambda x: "model_version_" in x, df.columns))
+            + list(filter(lambda x: "source_" in x, df.columns))
+            + list(filter(lambda x: "seed_" in x, df.columns))
+            + list(filter(lambda x: "prompt_embedding_" in x, df.columns))
+            + ["content_id", "guidance_scale", "num_inference_steps"]
+        )
+        content_features = df[content_columns]
+        content_features_tensor = torch.FloatTensor(content_features.values)
+        del content_features
+
+        DISLIKE_ENGAGEMENT_TYPE_VALUE = 0
+        LIKE_ENGAGEMENT_TYPE_VALUE = 1
+        MS_ENGAGEMENT_TYPE_VALUE = 2
+
+        engagement_type_value = df[["engagement_type", "engagement_value"]]
+        engagement_type_value["engagement_type"] = engagement_type_value.apply(
+            lambda x: (
+                LIKE_ENGAGEMENT_TYPE_VALUE
+                if x["engagement_type"] == "Like" and x["engagement_value"] == 1
+                else DISLIKE_ENGAGEMENT_TYPE_VALUE  # Like
+                if x["engagement_type"] == "Like" and x["engagement_value"] == -1
+                else MS_ENGAGEMENT_TYPE_VALUE  # Dislike  # MsEngagementType
+            ),
+            axis=1,
+        )
+
         user_tensor = df_to_user_tensor(df)
         if len(df["user_id"].unique()) != len(user_tensor):
             logging.error("Mismatch in user tensor length")
             return np.array([])
 
-        embeddings = self.model.forward_user(user_tensor).detach().numpy().astype(np.float32)
+        embeddings = (
+            self.model.forward_user(user_tensor).detach().numpy().astype(np.float32)
+        )
         if len(embeddings) != len(user_tensor) or embeddings.shape[1] > 64:
-            logging.error("Mismatch in embeddings and tensor length or embedding size exceeds 64")
+            logging.error(
+                "Mismatch in embeddings and tensor length or embedding size exceeds 64"
+            )
             return np.array([])
 
         return embeddings
 
+
 # Initialize ModelWrapper with an empty path to return a dummy model for testing
 model_wrapper = ModelWrapper()
```

Comment:
are you missing a path here?

Reply:
Hi Professor, we will edit it to remove that line of code:
model_wrapper = ModelWrapper()

Because the ModelWrapper class is instantiated separately in the two_tower_ann.py file where the path is specified